pilot/pkg/proxy/envoy/v2/gen2.go

```diff
@@ -0,0 +1,145 @@
+
+
+
+var (
+	// Interface is slightly different for now - LDS/CDS have extra param.
+	gen = &grpcgen.GrpcConfigGenerator{}
+)
+
+// handleReqAck checks if the message is an ack/nack and handles it, returning true.
+// If false, the request should be processed by calling the generator.
+func (s *DiscoveryServer) handleReqAck(con *XdsConnection, discReq *xdsapi.DiscoveryRequest) (*model.WatchedResource, bool) {
+
+	// All NACKs should have ErrorDetail set !
+	// Relying on versionCode != sentVersionCode as nack is less reliable.
+
+	isAck := true
+
+	t := discReq.TypeUrl
+	con.mu.RLock()
+	w := con.node.Active[t]
+	if w == nil {
+		w = &model.WatchedResource{
+			TypeUrl:  t,
+		}
+		con.node.Active[t] = w
+		isAck = false // newly watched resource
+	}
+	con.mu.RUnlock()
+
+	if discReq.ErrorDetail != nil {
+		errCode := codes.Code(discReq.ErrorDetail.Code)
+		adsLog.Warnf("ADS: ACK ERROR %s %s:%s", con.ConID, errCode.String(), discReq.ErrorDetail.GetMessage())
+		return w, true
+	}
+
+	if discReq.ResponseNonce == "" {
+		isAck = false // initial request
+	}
+	// This is an ACK response to a previous message - but it may refer to a response on a previous connection to
+	// a different XDS server instance.
+	nonceSent := w.NonceSent
+
+	// GRPC doesn't send version info in NACKs for RDS. Technically if nonce matches
+	// previous response, it is an ACK/NACK.
+	if nonceSent != "" && nonceSent == discReq.ResponseNonce {
+		adsLog.Debugf("ADS: ACK %s %s %s %v", con.ConID, discReq.VersionInfo, discReq.ResponseNonce,
+			time.Since(w.LastSent))
+		w.NonceAcked = discReq.ResponseNonce
+	}
+
+	if nonceSent != discReq.ResponseNonce {
+		adsLog.Debugf("ADS:RDS: Expired nonce received %s, sent %s, received %s",
+			con.ConID, nonceSent, discReq.ResponseNonce)
+		rdsExpiredNonce.Increment()
+		// This is an ACK for a resource sent on an older stream, or out of sync.
+		// Send a response back.
+		isAck = false
+	}
+
+	// Change in the set of watched resource - regardless of ack, send new data.
+	if !listEqualUnordered(w.ResourceNames, discReq.ResourceNames) {
+		isAck = false
+		w.ResourceNames = discReq.ResourceNames
+	}
+
+	return w, isAck
+}
+
+
+// handleCustomGenerator uses model.Generator to generate the response.
+func (s *DiscoveryServer) handleCustomGenerator(con *XdsConnection, req *xdsapi.DiscoveryRequest) error {
+	w, isAck := s.handleReqAck(con, req)
+	if isAck {
+		return nil
+	}
+
+	push := s.globalPushContext()
+	resp := &xdsapi.DiscoveryResponse{
+		TypeUrl:     w.TypeUrl,
+		VersionInfo: push.Version, // TODO: we can now generate per-type version !
+		Nonce:       nonce(push.Version),
+	}
+
+	cl := con.node.Generator.Generate(con.node, push, w)
+	for _, rc := range cl {
+		resp.Resources = append(resp.Resources, rc)
+	}
+
+	err := con.send(resp)
+	if err != nil {
+		adsLog.Warnf("ADS: Send failure %s: %v", con.ConID, err)
+		recordSendError(apiSendErrPushes, err)
+		return err
+	}
+	w.LastSent = time.Now()
+	w.LastSize = proto.Size(resp)
```

Comment:
nit: I think this may have tangible perf impact

Reply:
Yes, I know - added a comment. Someone asked about the sizes - but better to do it on client. 
I think I can just use the Any.Body since it's a byte[]

Reply:
Fixed to use the resource size only.